docs/architecture/adr-062-p2p-architecture.md

```diff
@@ -0,0 +1,531 @@
+
+The `peer` struct might look like the following, but is intentionally underspecified and will depend on implementation requirements (for example, it will almost certainly have to track statistics about connection failures and retries):
+
+```go
+// peer tracks internal status information about a peer.
+type peer struct {
+    ID        PeerID
+    Status    PeerStatus
+    Priority  PeerPriority
+    Endpoints map[PeerAddress][]Endpoint // Resolved endpoints by address.
+}
+
+// PeerStatus specifies peer statuses.
+type PeerStatus string
+
+const (
+    PeerStatusNew     = "new"     // New peer which we haven't tried to contact yet.
+    PeerStatusUp      = "up"      // Peer which we have an active connection to.
+    PeerStatusDown    = "down"    // Peer which we're temporarily disconnected from.
+    PeerStatusRemoved = "removed" // Peer which has been removed.
+    PeerStatusBanned  = "banned"  // Peer which is banned for misbehavior.
+)
+
+// PeerPriority specifies peer priorities.
+type PeerPriority int
+
+const (
+    PeerPriorityNormal PeerPriority = iota + 1
+    PeerPriorityValidator
+    PeerPriorityPersistent
+)
+```
+
+Peer information is stored in a `peerStore`, which may be persisted in an underlying database, and will replace the current address book either partially or in full. It is kept internal to avoid race conditions and tight coupling, and should at the very least contain basic CRUD functionality as outlined below (but will likely need additional functionality and is intentionally underspecified):
+
+```go
+// peerStore contains information about peers, possibly persisted to disk.
+type peerStore struct {
+    peers map[string]*peer // Entire set in memory, with PeerID.String() keys.
+    db    dbm.DB           // Database for persistence, if non-nil.
+}
+
+func (p *peerStore) Delete(id PeerID) error     { return nil }
+func (p *peerStore) Get(id PeerID) (peer, bool) { return peer{}, false }
+func (p *peerStore) List() []peer               { return nil }
+func (p *peerStore) Set(peer peer) error        { return nil }
+```
+
+### Channels
+
+While low-level data exchange happens via transport IO streams, the high-level API is based on a bidirectional `Channel` that can send and receive Protobuf messages addressed by `PeerID`. A channel is identified by an arbitrary `ChannelID` identifier, and can exchange Protobuf messages of one specific type (since the type to unmarshal into must be known). Message delivery is asynchronous and at-most-once.
+
+A `Channel` has this interface:
+
+```go
+// Channel is a bidirectional channel for Protobuf message exchange with peers.
+type Channel struct {
+    // ID contains the channel ID.
+    ID ChannelID
+
+    // messageType specifies the type of messages exchanged via the channel, and
+    // is used e.g. for automatic unmarshaling.
+    messageType proto.Message
+
+    // In is a channel for receiving inbound messages. Envelope.From is always
+    // set.
+    In <-chan Envelope
```

Comment:
Will this be buffered? What happens when buffer is full?

Reply:
Yes. This is mentioned elsewhere - message delivery is asynchronous and at-most-once, and will be buffered, dropped, or blocked depending on the scheduling policy of the router. This is by design, and is one mechanism to do backpressure when peers are overloaded or slow.